src/day9.scala

```diff
@@ -0,0 +1,112 @@
+// using scala 3.1.0
+
+package day9
+
+import scala.collection.immutable.Queue
+import scala.util.Using
+import scala.io.Source
+
+@main def part1(): Unit =
+  println(s"The solution is ${part1(readInput())}")
+
+@main def part2(): Unit =
+  println(s"The solution is ${part2(readInput())}")
+
+def readInput(): String =
+  Using.resource(Source.fromFile("input/day9"))(_.mkString)
+
+type Height = Int
+case class Position(x: Int, y: Int)
+
+class Heightmap(val width: Int, val height: Int, private val data: Array[Height]):
+
+  def apply(coord: Position): Height = data(coord.y * width + coord.x)
+
+  def neighborsOf(coord: Position): Iterator[(Position, Height)] =
+    val Position(x, y) = coord
+    val idx = coord.y * width + coord.x
+    Iterator(
+      Option.when(x > 0)(Position(x - 1, y) -> data(idx - 1)),
+      Option.when(x < width - 1)(Position(x + 1, y) -> data(idx + 1)),
+      Option.when(y > 0)(Position(x, y - 1) -> data(idx - width)),
+      Option.when(y < height - 1)(Position(x, y + 1) -> data(idx + width))
+    ).flatMap(Iterator.from)
```

Comment:
Here you're directly accessing `data` with flat indices, even though you defined an abstraction with `Position` as `apply`. You may want to reuse it. Perhaps like this:
```suggestion
    Iterator(
      Option.when(x > 0)(Position(x - 1, y)),
      Option.when(x < width - 1)(Position(x + 1, y)),
      Option.when(y > 0)(Position(x, y - 1)),
      Option.when(y < height - 1)(Position(x, y + 1))
    ).flatMap(Iterator.from).map(coord => coord -> apply(coord))
```